add codeparrot blog post (#184)
huggingface/blog: codeparrot.md

---
title: Training CodeParrot 🦜 from Scratch
thumbnail: /blog/assets/40_codeparrot/thumbnail.png
---

<h1>Training CodeParrot 🦜 from Scratch</h1>

<div class="blog-metadata">
    <small>Published Dec 8, 2021.</small>
    <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/master/codeparrot.md">
        Update on GitHub
    </a>
</div>

<div class="author-card">
    <a href="/leandro">
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1627890220261-5e48005437cb5b49818287a5.png?w=200&h=200&f=face" title="Gravatar">
        <div class="bfc">
            <code>lvwerra</code>
            <span class="fullname">Leandro von Werra</span>
        </div>
    </a>
</div>

> Let's create all the code for training GitHub CoPilot in this blog post. Learn how to train CodeParrot 🦜, a large GPT-2 model for code, from scratch in this step by step guide. Let's get to it!

Reviewer: Nit

```suggestion
Let's create all the code for training GitHub Copilot in this blog post. Learn how to train CodeParrot 🦜, a large GPT-2 model for code, from scratch in this step by step guide. Let's get to it!
```

Reviewer: Maybe we can kick things off for an audience that is a bit less technical or an intro that has a stronger hook.

> Would you like to train a ML model that generates code? In this blog post we'll see how to train CodeParrot 🦜, a large code generation model from scratch. Let's do it.

Actually not convinced by this but you got what I meant :D

> ![codeparrot](assets/40_codeparrot/codeparrot.png)

Reviewer: Thanks for using small images!

Author: Is that a comment or request? I made the image a bit smaller.

Reviewer: That was an appreciation comment :D the current sizes are great. Sometimes people put multiple MB images in the blog which is not great from user perspective.

> ## Dataset

Reviewer 2: Suggestion:

```suggestion
## Creating a Large Dataset of Source Code
```

> The first thing we need is a large training dataset. With the goal to train a Python code generation model we accessed the GitHub dump available on Google's BigQuery and filtered for all Python files. The result is a 180 GB dataset with 20 million files (available [here](http://hf.co/datasets/transformersbook/codeparrot)). After initial training experiments we found that the duplicates in the dataset severly impacted the model performance. Further investigating the dataset we found that:

Reviewer: ```suggestion
The first thing we need is a large training dataset. With the goal to train a Python code generation model, we accessed the GitHub dump available on Google's BigQuery and filtered for all Python files. The result is a 180 GB dataset with 20 million files (available [here](http://hf.co/datasets/transformersbook/codeparrot)). After initial training experiments, we found that the duplicates in the dataset severely impacted the model performance. Further investigating the dataset we found that:
```

- 0.1% of the unique files make up 15% of all files
- 1% of the unique files make up 35% of all files

> - 10% of the unique files make up 66% of all files

Reviewer: Is there any chart to visualize this?

Author: I can make one if you think it is important but I don't have one readily available :) I added a link to the Twitter thread instead.

> We removed the duplicates and applied the same cleaning heuristics found in the [Codex paper](https://arxiv.org/abs/2107.03374). The cleaned dataset is still 50GB big and available on the 🤗 Hub: [codeparrot-clean](http://hf.co/datasets/lvwerra/codeparrot-clean). With that we can setup a new tokenizer and train a model model.

Reviewer: ```suggestion
We removed the duplicates and applied the same cleaning heuristics found in the [Codex paper](https://arxiv.org/abs/2107.03374). The cleaned dataset is still 50GB big and available on the Hugging Face Hub: [codeparrot-clean](http://hf.co/datasets/lvwerra/codeparrot-clean). With that we can setup a new tokenizer and train a model model.
```

> 

Reviewer: Is the script to pre-process the data available?

> ## Initializing Tokenizer and Model

Reviewer 2: nit:

```suggestion
## Initializing the Tokenizer and Model
```

> First we need a tokenizer. Let's train one specifically on code so it splits code tokens well. We can take an existing tokenizer (e.g. GPT-2) and directly train it on our own dataset with `train_new_from_iterator`. We then push it to the hub.

Reviewer: ```suggestion
First we need a tokenizer. Let's train one specifically on code so it splits code tokens well. We can take an existing tokenizer (e.g. GPT-2) and directly train it on our own dataset with `train_new_from_iterator`. We then push it to the Hub.
```

> 

Reviewer: Maybe let's add a link to the course section of tokenizers somewhere in this section. Somewhat like "Is this the first time you read about training your own tokenizers? Don't worry! We got you covered. ..."

> ```Python

Reviewer 2: Does this produce the correct color syntax highlighting or should it be `python`? Not sure if it matters or not ...

# Iterator for Training
def batch_iterator(batch_size=10):
    for _ in tqdm(range(0, args.n_examples, batch_size)):
        yield [next(iter_dataset)["content"] for _ in range(batch_size)]

# Base tokenizer
tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
base_vocab = list(bytes_to_unicode().values())

# Load dataset
dataset = load_dataset("lvwerra/codeparrot-clean", split="train", streaming=True)
iter_dataset = iter(dataset)

# Training and saving
new_tokenizer = tokenizer.train_new_from_iterator(batch_iterator(),
                                                  vocab_size=args.vocab_size,
                                                  initial_alphabet=base_vocab)
new_tokenizer.save_pretrained(args.tokenizer_name, push_to_hub=args.push_to_hub)
```

Now, we initialize a new model. We’ll use the same hyperparameters as GPT-2 large (1.5B parameters) and adjust the embedding layer to fit our new tokenizer also adding some stability tweaks. We push the freshly initialized model to the same repo as the tokenizer.

```Python
# Load codeparrot tokenizer trained for Python code tokenization
tokenizer = AutoTokenizer.from_pretrained(args.tokenizer_name)

> # Config: "scale_attn_by_layer_idx" and "reorder_and_upcast_attn" are  stability tweaks

Reviewer 2: I think it would be nice to add a sentence or two that explains what is meant by stability tweaks. I think this is a really interesting subject and you could link to e.g. Stas' tweets on BigScience training

config_kwargs = {"vocab_size": len(tokenizer),
                 "scale_attn_by_layer_idx": True,
                 "reorder_and_upcast_attn": True}

# Load model with config and push to hub
config = AutoConfig.from_pretrained(gpt2-large, **config_kwargs)
model = AutoModelForCausalLM.from_config(config)
model.save_pretrained(args.model_name, push_to_hub=args.push_to_hub)
```

Now that we have an efficient tokenizer and a freshly initialized model we can start with the actual training loop.

> ## Training Loop

Reviewer 2: Suggestion: 

```suggestion
## Implementing the Training Loop
```

> We train with the 🤗 Accelerate library which allows us to scale the training from our laptop to a multi-GPU machine without changing a single line of code. We just create an accelerator and do some argument housekeeping:

Reviewer: ```suggestion
We train with the [🤗 Accelerate](https://github.com/huggingface/accelerate) library which allows us to scale the training from our laptop to a multi-GPU machine without changing a single line of code. We just create an accelerator and do some argument housekeeping:
```

```Python
accelerator = Accelerator()
acc_state = {str(k): str(v) for k, v in accelerator.state.__dict__.items()}

parser = HfArgumentParser(TrainingArguments)
args = parser.parse_args()
args = Namespace(**vars(args), **acc_state)
samples_per_step = accelerator.state.num_processes * args.train_batch_size
set_seed(args.seed)
```

> We are now ready to train. Let's use the hub library to clone the repository with the new tokenizer and model and we checkout a new branch for this experiment. With that setup we can run many experiments in parallel and in the end we just merge the best one into the main branch.

Reviewer: ```suggestion
We are now ready to train! Let's use the `huggingface_hub` client library to clone the repository with the new tokenizer and model. We will checkout to a new branch for this experiment. With that setup, we can run many experiments in parallel and in the end we just merge the best one into the main branch.
```

> We are now ready to train. Let's use the hub library to clone the repository with the new tokenizer and model and we checkout a new branch for this experiment. With that setup we can run many experiments in parallel and in the end we just merge the best one into the main branch.

Reviewer: Maybe worth adding a section with pip installs beforehand

```Python
# Clone model repository
if accelerator.is_main_process:

>     hf_repo = Repository(args.save_dir, clone_from=args.model_ckpt)

Reviewer: Should we also add the import to `Repository`?

# Checkout new branch on repo
if accelerator.is_main_process:
    hf_repo.git_checkout(run_name, create_branch_ok=True)
```

> We can directly load the tokenizer and model from the local repository. Since we are dealing with big models we might want to turn on gradient checkpointing to decrease the GPU memory footprint during training.

Reviewer: Maybe let's add a link to what gradient checkpointing is.

```Python
# Load model and tokenizer
model = AutoModelForCausalLM.from_pretrained(args.save_dir)
if args.gradient_checkpointing:
    model.gradient_checkpointing_enable()
tokenizer = AutoTokenizer.from_pretrained(args.save_dir)
```

Next up is the dataset. We make training simpler with a dataset that yields examples with a fixed context size. To not waste too much data (some samples are too short or too long) we can concatenate many examples with an EOS token and then chunk them.

![codeparrot](assets/40_codeparrot/buffer.png)

> The more sequences we prepare together, the smaller the fraction of tokens we discard (the grey ones in the previous figure). Since we want to stream the dataset instead of preparing everything in advance we use an `IterableDataset`. The full dataset class looks as follows:

Reviewer: Maybe worth adding one sentence of what streaming is before the second sentence

Reviewer 2: You could also link to the course [chapter](https://huggingface.co/course/chapter5/4?fw=pt) or just the video on [streaming](https://huggingface.co/course/chapter5/4?fw=pt) if you want

Author: good point, i had them in the thread but they did not make it to the blog for some reason.

        more_examples = True
        while more_examples:
            buffer, buffer_len = [], 0
            while True:
                if buffer_len >= self.input_characters:
                    break
                try:
                    buffer.append(next(iterator)["content"])
                    buffer_len += len(buffer[-1])
                except StopIteration:
                    if self.infinite:
                        iterator = iter(self.dataset)
                        self.epoch += 1
                        logger.info(f"Dataset epoch: {self.epoch}")
                    else:
                        more_examples = False
                        break
            tokenized_inputs = self.tokenizer(buffer, truncation=False)["input_ids"]
            all_token_ids = []
            for tokenized_input in tokenized_inputs:
                all_token_ids.extend(tokenized_input + [self.concat_token_id])
            for i in range(0, len(all_token_ids), self.seq_length):
                input_ids = all_token_ids[i : i + self.seq_length]
                if len(input_ids) == self.seq_length:
                    yield torch.tensor(input_ids)
```

> Here texts in the buffer are tokenized in parallel and then concatenated. Chunked samples are then yielded until the buffer is empty and the process starts again. If we set infinite=True the dataset iterator restarts at its end.

Reviewer: ```suggestion
Texts in the buffer are tokenized in parallel and then concatenated. Chunked samples are then yielded until the buffer is empty and the process starts again. If we set infinite=True the dataset iterator restarts at its end.
```

        if any(nd in n for nd in no_decay): params_without_wd.append(p)
        else: params_with_wd.append(p)
    return [{"params": params_with_wd, "weight_decay": args.weight_decay},
            {"params": params_without_wd, "weight_decay": 0.0},]

optimizer = AdamW(get_grouped_params(model, args), lr=args.learning_rate)
lr_scheduler = get_scheduler(name=args.lr_scheduler_type, optimizer=optimizer,
                             num_warmup_steps=args.num_warmup_steps,
                             num_training_steps=args.max_train_steps,)
```

A big question that remains is how all the data and models will be distributed across several GPUs. This sounds like a complex task but actually only requires a single line of code with 🤗 Accelerate.

```Python
model, optimizer, train_dataloader, eval_dataloader = accelerator.prepare(
    model, optimizer, train_dataloader, eval_dataloader)
```
Under the hood it'll use DistributedDataParallel, which means a batch is sent to each GPU worker which has its own copy of the model. There the gradients are computed and then aggregated to update the model on each worker.

![codeparrot](assets/40_codeparrot/ddp.png)

> We also want to evaluate the model from time to time on the validation set so let’s write a function to do just that. This is done automatically in a distributed fashion and we just need to gather all the losses from the workers. We also want to report the perplexity.

Reviewer: Maybe worth adding a link for the perplexity subsection in the course

```Python
# Train model
model.train()
completed_steps = 0
for step, batch in enumerate(train_dataloader, start=1):
    loss = model(batch, labels=batch, use_cache=False).loss
    loss = loss / args.gradient_accumulation_steps
    accelerator.backward(loss)
    if step % args.gradient_accumulation_steps == 0:
        accelerator.clip_grad_norm_(model.parameters(), 1.0)
        optimizer.step()
        lr_scheduler.step()
        optimizer.zero_grad()
        completed_steps += 1
    if step % args.save_checkpoint_steps == 0:
        eval_loss, perplexity = evaluate(args)
        accelerator.wait_for_everyone()
        unwrapped_model = accelerator.unwrap_model(model)
        unwrapped_model.save_pretrained(args.save_dir, save_function=accelerator.save)
        if accelerator.is_main_process:
            hf_repo.push_to_hub(commit_message=f"step {step}")
        model.train()
    if completed_steps >= args.max_train_steps:
        break
```

> When we call `wait_for_everyone` and `unwrap` we make sure that all workers are ready and any model layers that have been added by `prepare` earlier are removed. We also use gradient accumulation and gradient clipping that are easily implemented. Lastly, after training is complete we run a last evaluation and save the final model and push it to the hub. 

Reviewer 2: nit:

```suggestion
When we call `wait_for_everyone()` and `unwrap_model()` we make sure that all workers are ready and any model layers that have been added by `prepare()` earlier are removed. We also use gradient accumulation and gradient clipping that are easily implemented. Lastly, after training is complete we run a last evaluation and save the final model and push it to the hub. 
```

```Python
# Evaluate and save the last checkpoint
logger.info("Evaluating and saving model after training")
eval_loss, perplexity = evaluate(args)
log_metrics(step, {"loss/eval": eval_loss, "perplexity": perplexity})
accelerator.wait_for_everyone()
unwrapped_model = accelerator.unwrap_model(model)
unwrapped_model.save_pretrained(args.save_dir, save_function=accelerator.save)
if accelerator.is_main_process:
    hf_repo.push_to_hub(commit_message="final model")
```
Done! That's all the code to train a full GPT-2 model from scratch with as little as 150 lines. We did not show the imports and logs of the scripts to make the code a little bit more compact. But you'll find the full code including preprocessing and downstream task evaluation [here](https://github.com/huggingface/transformers/tree/master/examples/research_projects/codeparrot). Now let's actually train it!

> With this code we trained models for the upcoming [transformers book](https://learning.oreilly.com/library/view/natural-language-processing/9781098103231/): a [110M](https://hf.co/lvwerra/codeparrot-small) and [1.5B](https://hf.co/lvwerra/codeparrot) parameter GPT-2 model. We used a 16 x A100 GPU machine to train these models for 1 day and 1 week, respectively. Enough time to get a coffee and read a book or two!

Reviewer 2: Suggestion: 

```suggestion
With this code we trained models for our upcoming [book on Transformers and NLP](https://learning.oreilly.com/library/view/natural-language-processing/9781098103231/): a [110M](https://hf.co/lvwerra/codeparrot-small) and [1.5B](https://hf.co/lvwerra/codeparrot) parameter GPT-2 model. We used a 16 x A100 GPU machine to train these models for 1 day and 1 week, respectively. Enough time to get a coffee and read a book or two!
```

## Evaluation

> This is still relatively short training time for pretraining but we can already observe good downstream performance as compared to similar models. We evaluated the models on OpenAI's HumanEval benchmark that was introduced in the [Codex paper](https://arxiv.org/abs/2107.03374).

Reviewer 2: Maybe mention that you can access this benchmark and evaluate models on it with `datasets`?

![codeparrot](assets/40_codeparrot/humaneval.png)

Note that we trained CodeParrot on roughly 25-30B tokens whereas GPT-neo was trained on 300B tokens and Codex on 300B (GPT-3 checkpoint) + 100B (code) tokens. So you definitely get a good bang for your buck here.

## Demos

> You can play with the models in two demos we added to 🤗 Spaces. With the first you can quickly generate code with the model and with the second you can highlight your code with the model to spot bugs!

Reviewer: ```suggestion
You can play with the models in two demos we added to [Hugging Face Spaces](https://huggingface.co/spaces/launch). With the first you can quickly generate code with the model and with the second you can highlight your code with the model to spot bugs!
```

- [Code Generation](https://hf.co/spaces/lvwerra/codeparrot-generation)
- [Code Highlighting](https://hf.co/spaces/lvwerra/codeparrot-highlighting)

> 

Reviewer: Maybe you could have a short results section that show 2 or 3 interesting results?
